synapse/rest/media/v1/upload_resource.py

```diff
@@ -104,3 +118,42 @@ async def _async_render_POST(self, request: SynapseRequest) -> None:
         logger.info("Uploaded content with URI %r", content_uri)
 
         respond_with_json(request, 200, {"content_uri": content_uri}, send_cors=True)
+
+    async def _async_render_PUT(self, request: SynapseRequest) -> None:
+        if not self.enable_async_uploads:
+            raise SynapseError(
+                405,
+                "Asynchronous uploads are not enabled on this homeserver",
+                errcode=Codes.UNRECOGNIZED,
+            )
+
+        requester = await self.auth.get_user_by_req(request)
+        server_name, media_id, _ = parse_media_id(request)
+
+        if server_name != self.server_name:
+            raise SynapseError(
+                404,
+                "Non-local server name specified",
+                errcode=Codes.NOT_FOUND,
+            )
+
+        await self.media_repo.verify_can_upload(media_id, requester.user)
+        content_length, upload_name, media_type = self._get_file_metadata(request)
+
+        try:
+            content: IO = request.content  # type: ignore
+            await self.media_repo.update_content(
+                media_id,
+                media_type,
+                upload_name,
+                content,
+                content_length,
+                requester.user,
+            )
+        except SpamMediaException:
+            # For uploading of media we want to respond with a 400, instead of
+            # the default 404, as that would just be confusing.
+            raise SynapseError(400, "Bad content")
```

Comment:
This is the section that I am referring to in the description of this PR regarding a potential race condition between verifying and updating the content of the media ID.

Reply:
As I understand it, the code does something like:
```
if media does not have content:
    write file
    update db row
```
And the spec mandates that the content can only be set once, so if multiple requests to this endpoint come in concurrently, we violate the spec?

Reply:
That is correct. According to: https://github.com/matrix-org/matrix-spec-proposals/pull/2246/files#diff-df761ec989cb82d0a79b226b96070b7da7f995ef6093f5a6ca4a6817cfb941e8R46-R48

```
Upload content to a MXC URI that was created earlier. If the endpoint is called
with a media ID that already has content, the request should be rejected with
the error code `M_CANNOT_OVERWRITE_MEDIA` and HTTP status code 409.
```

Reply:
One solution could be to have per-worker locks (using `Linearizer` maybe?) on media IDs and rely on admins to configure haproxy/nginx to always route requests to do with the same media ID to the same worker. Personally I don't find this very satisfactory though it may be the simplest solution?

An alternative might involve file locks (though I'm not familiar with them). eg.
```
lock the file to prevent other writes (or block)
    check that someone else didn't update the db
    write file
    update db row
unlock the file
```

Reply:
Resolved by using `LockStore.try_acquire_lock` with the changes from https://github.com/matrix-org/synapse/pull/12832